### DISCUSSION TITLE
Ordered Lists, Callouts and Ignoring Nodes

### DISCUSSION THREAD

**Discussion by celtic-coder** [pml-lang/pml-companion#30] [Ideas]
Hi Christian (@pml-lang),

While working on a document this past weekend, I thought about three possible enhancements that I would like to suggest for your consideration. In the grand scheme of things, they might be classed as "nice to have" or low priority, but it might be useful to keep in mind in any case:

### Ordered Lists

I notice that the default `[list ... ]` tag produces the unordered list (`<ul>...</ul>`) of elements. Might there be an enhancement to produce ordered lists at some stage in the future?

I tried two different workarounds:

(1) I copied the HTML code produced for the unordered list, but substituted the `<ol>...</ol>` tags instead. The `[html ... html]` for everything in the list worked fine, but it might be cumbersome:

![html-tags](https://user-images.githubusercontent.com/11787866/116295549-1d622580-a791-11eb-9bbe-9f4c33e70c33.png)

(2) The best method that I could come up with, was a combination of `[html...html]` and the `[el...]` tags:

![html-ol-ol-tags](https://user-images.githubusercontent.com/11787866/116296218-db85af00-a791-11eb-8bb9-54d540798802.png)

This is a bit of a "hack", but it is a reasonable workaround for when an ordered list is required.

### Callouts

Your description of the Simplix language uses callouts to highlight details in the code. For example, to explain the steps in the "[how to count lines of code](https://www.codeproject.com/Articles/1244412/How-to-Design-and-Build-the-Simplest-Possible-Prog#count_lines_of_code)" snippet:

![simplex-callouts](https://user-images.githubusercontent.com/11787866/116321195-82c60e80-a7b1-11eb-8983-61ad5d00a622.png)

I have seldom used callouts, but when it is required it is useful. It is one of the features that I like about AsciiDoctor. Will you be incorporating this functionality into PML at some stage?

### Ignoring Nodes in Tree

In the [Sublime PML Discussion](https://github.com/tajmone/Sublime-PML/discussions/14#discussioncomment-653318), I had mentioned that I use the Leo Editor to write the outline and the content for my documents.  One of its features in writing AsciiDoctor, for example, is that nodes in the outline can be [ignored](https://github.com/leo-editor/leo-editor/blob/e46e9315f61a1c378900d3ba7fbadf39c5d9fb5b/leo/core/leoMarkup.py#L33-L41):

I have used this in the past to create a node (and its children), that contain background information for display during the writing the document, but is ignored once the text is complete. It is useful for the writing stage where the outline is chopped and changed and material is moved around as ideas are constantly modified. Nodes several levels deep can be ignored without affected the parent nodes.

I realize that this is a "nice-to-have" since I am so used to an outlining editor and PML is designed to be a more static writing instrument. However, it might be the germ of an idea that might prove useful at some stage in the future.

Kind Regards,
Liam

**Comment by pml-lang** [comment]
> Might there be an enhancement to produce ordered lists at some stage in the future?

Yes, this is already on my todo list, as a _must-have_ feature. Maybe the syntax will be `[list type=ordered`, but not sure yet. I will soon have more time to work on PML again, and this is surely one of the new features to appear in an upcoming version.

> I have seldom used callouts, but when it is required it is useful. It is one of the features that I like about AsciiDoctor. Will you be incorporating this functionality into PML at some stage?

Callouts are sometimes very useful, indeed, and they look good. It's currently classified as a 'nice to have' on my list. If you need it more quickly, please just let me know.

> nodes in the outline can be ignored

Whenever I want to hide nodes in my PML documents (or text like a TODO), I embed them in a comment, so that they will be ignored by the converter. I never used Leo, but I can image that a dedicated outline editor is much more convenient and powerful for manipulating nodes. I wonder: Would it be possible to add PML support in Leo?

Thanks for your suggestions, Liam.

**Reply by pml-lang** [reply]
Update:

> Might there be an enhancement to produce ordered lists at some stage in the future?

I just wanted to implement `[list type=ordered`, when I realized that it's not necessary. You can have an ordered (numbered) list like this:
```
Numbered list:
[list html_style="list-style-type:decimal"
    [el item 1]
    [el item 2]
]
```
I have updated the [reference manual](https://www.pml-lang.dev/docs/reference_manual/index.html#node_list) (changed sections 'Description' and 'Examples'). Please have a look at section 'Examples' for more information.

**Comment by celtic-coder** [comment]
Hi Christian (@pml-lang),

Leo is a Python application and it has a number of importers / exporters for different languages. It has native support for reStructured Text and AsciiDoctor so, I presume, that it could handle PML also. It might need some dedicated work to get PML work with the rest of the ecosystem. One developer has used Vue.js to implement an interesting [viewer](https://kaleguy.github.io/leovue) for Leo outlines. The page has a useful summary of Leo's features, but also an extensive outline showing its capabilities.

Yes, commenting out a node is the most straightforward method of removing a node from the final output.

Kind Regards,
Liam

**Reply by pml-lang** [reply]
LeoVue looks good and has an intriguing set of features.
Thanks a lot for the link.
